### PR TITLE
Model Cache Management Features

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
features

* **What is the current behavior?** (You can also link to an open issue here)
users can not view the loaded models and eject a model from the client

* **What is the new behavior (if this is a feature change)?**
users can now view loaded models, eject models, view cuda information from the client

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
no

* **Have unit tests been run against this PR?** (Has there also been any additional testing?)
not yet

* **Related Python client changes** (link commit/PR here)
will link later

* **Related documentation changes** (link commit/PR here)
will link later

* **Other information**:
this PR is related to the [issue](https://github.com/marqo-ai/marqo/issues/204#issuecomment-1352574646)

* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by pandu-k** [comment]
> Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)

FYI 
Adding new API endpoints is not a breaking change. A breaking change is updating an existing endpoint such that it is no longer backwards compatible (so existing users have to update their code to use the new version).

An example of a breaking change is when the add_documents `batch_size` parameter was replaced by `server_batch_size` and `client_batch_size` parameters. Users wanting to use the new py-marqo version would have to update their add_documents calls

**Comment by wanliAlex** [comment]
> 

Thanks for the comment.

File: src/marqo/tensor_search/api.py [hunk 1]
```diff
@@ -212,6 +212,16 @@ def check_health(marqo_config: config.Config = Depends(generate_config)):
...
+@app.get("/models")
+def get_loaded_models():
+    return tensor_search.get_loaded_models()
+@app.delete("/models")
+def eject_model(model_name:str, model_device:str):
+    return tensor_search.eject_model(model_name = model_name, device = model_device)
+@app.get("/device/cuda")
+def get_cuda_info():
+    return tensor_search.get_cuda_info()
+
```
Comment (pandu-k) [inline src/marqo/tensor_search/api.py:234]: Double spaces between functions please

File: src/marqo/errors.py [hunk 2]
```diff
@@ -176,6 +176,10 @@ class IndexMaxFieldsError(__InvalidRequestError):
     code = "index_max_fields_error"
     status_code = HTTPStatus.BAD_REQUEST
 
+class ModelNotLoadedError(__InvalidRequestError):
+    code = "model_not_loaded"
+    status_code = HTTPStatus.NOT_FOUND
+
```
Comment (pandu-k) [inline src/marqo/errors.py:183]: Would ModelNotInCache error be better? 

ModelNotLoaded is ambiguous on whether it is about an action that failed ('loading failed') or a state ('model not found in cache').

File: src/marqo/s2_inference/errors.py [hunk 3]
```diff
@@ -42,3 +42,6 @@ class RerankerImageError(S2InferenceError):
 
 class RerankerNameError(S2InferenceError):
     pass
+
+class ModelNotLoadedError(S2InferenceError):
+    pass
```
Comment (pandu-k) [inline src/marqo/s2_inference/errors.py:48]: Would ModelNotInCache error be better?

ModelNotLoaded is ambiguous on whether it is about an action that failed ('loading failed') or a state ('model not found in cache').

File: src/marqo/s2_inference/s2_inference.py [hunk 4]
```diff
@@ -291,6 +292,19 @@ def _load_model(model_name: str, model_properties: dict, device: str = get_defau
 
     return model
 
+def get_available_models():
+    return available_models
+
```
Comment (pandu-k) [inline src/marqo/s2_inference/s2_inference.py:300]: double spaces between functions

File: src/marqo/s2_inference/s2_inference.py [hunk 5]
```diff
@@ -291,6 +292,19 @@ def _load_model(model_name: str, model_properties: dict, device: str = get_defau
...
+def eject_model(model_name:str,device:str):
+    model_cache_key = _create_model_cache_key(model_name, device)
+    if model_cache_key in available_models:
+        del available_models[model_cache_key]
+        if device.startswith("cuda"):
+            torch.cuda.empty_cache()
```
Comment (pandu-k) [inline src/marqo/s2_inference/s2_inference.py:321]: Does this empty the entire CUDA cache? We only want to eject the specified model. 

Also, how do specific CUDA devices (e.g.: `cuda:1`, just `cuda`) affect this, if at all?

File: src/marqo/tensor_search/api.py (hunk 1 above)
Comment (pandu-k) [reply src/marqo/tensor_search/api.py:234]: @jn2clark Thoughts on the device endpoint? How will this work with future possibilities like an inference server, and non-cuda APIs

File: src/marqo/tensor_search/tensor_search.py [hunk 6]
```diff
@@ -1238,3 +1239,22 @@ def _get_model_properties(index_info):
...
+def eject_model(model_name: str, device: str) -> dict:
+    try:
+       result = s2_inference.eject_model(model_name, device)
+    except s2_inference_errors.ModelNotLoadedError as e:
+        raise errors.ModelNotLoadedError(message=str(e))
+    return result
+def get_cuda_info() -> dict:
+    return {"device": "cuda",
+            "memory_usage": f"{round(torch.cuda.memory_allocated() / 1024**3, 1)} GiB",
+            "total_device_memory": f"{round(torch.cuda.get_device_properties(0).total_memory/ 1024**3, 1)} GiB"}
```
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py]: Double spaces between function please

Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py]: In this call `torch.cuda.get_device_properties(0)` what does the 0 refer to? Is that the index of the device (for example `cuda:0`)?

Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py]: The cuda functions should be moved to a separate file. I am thinking 
`s2_inference/devices/cuda.py`. We want to remain open to other designs.  ~~such as an inference server.~~

~~Thoughts @jn2clark?~~

**Review by pandu-k — CHANGES_REQUESTED** [review]
Thanks for this PR! Left actionable feedback in the comments.

Further to dos:
- Action feedback
- Create a series of unit tests. One that I propose is to load and unload a model ~10 times, to see if the cache is actually being unloaded (else the computer runs out of memory). Please include testing for unusual and edge cases. 
- Create corresponding Py-Marqo methods and a few unit tests
- Given this is a  new endpoint, create a few [API tests](https://github.com/marqo-ai/marqo-api-tests/tree/mainline/tests). These don't have to be as exhaustive as the unit tests, but should test every endpoint. 
  - An CUDA API test is proposed: load 2 CUDA models. Unload one. Test to ensure the remaining model is still cached. Test to ensure the unloaded model is indeed unloaded. Create API tests [here](https://github.com/marqo-ai/marqo-api-tests/tree/mainline/tests/api_tests/cuda). CUDA test classes are marked by [this decorator](https://github.com/marqo-ai/marqo-api-tests/blob/b909f35a6de35411340ed427af790d23c5c78643/tests/api_tests/cuda/test_cuda_add_documents.py#L12). This ensures that the API test only runs on the CUDA test environment.

File: src/marqo/tensor_search/api.py (hunk 1 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/api.py:234]: > Double spaces between functions please

solved

File: src/marqo/errors.py (hunk 2 above)
Comment (wanliAlex) [reply src/marqo/errors.py:183]: solved

File: src/marqo/s2_inference/errors.py (hunk 3 above)
Comment (wanliAlex) [reply src/marqo/s2_inference/errors.py:48]: sovled

File: src/marqo/s2_inference/s2_inference.py (hunk 4 above)
Comment (wanliAlex) [reply src/marqo/s2_inference/s2_inference.py:300]: solved

File: src/marqo/s2_inference/s2_inference.py (hunk 5 above)
Comment (wanliAlex) [reply src/marqo/s2_inference/s2_inference.py:321]: This will release the cached memory for other applications. 
If we don't call torch.cuda.empty.cached(), the memory will still be reserved for torch use. 
Adding this will also make the results consistent with `nvidia-smi`
This function does not take any argument.

File: src/marqo/tensor_search/tensor_search.py (hunk 6 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py]: solved

File: src/marqo/s2_inference/s2_inference.py (hunk 5 above)
Comment (pandu-k) [reply src/marqo/s2_inference/s2_inference.py:321]: Makes sense. I've left a note in the review notes about making an API test for this.

**Review by pandu-k — COMMENTED** [review]
There are still a few pending questions. Tests are also to be made. 

I have updated my [review comment](https://github.com/marqo-ai/marqo/pull/239#pullrequestreview-1222184138)

Resolved most conversations.

File: src/marqo/tensor_search/tensor_search.py (hunk 6 above)
Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py]: We now support multi-gpu information:

On 1-gpu device instance:
{'cuda_usage_info': [{'device_id': 0, 'device_name': 'Tesla V100-SXM2-16GB', 'memory_used': '0.0 GiB', 'total_memory': '15.8 GiB'}]}
​
On 4-gpu devices instance:
{'cuda_usage_info': [{'device_id': 0, 'device_name': 'Tesla V100-SXM2-16GB', 'memory_used': '0.0 GiB', 'total_memory': '15.8 GiB'}, {'device_id': 1, 'device_name': 'Tesla V100-SXM2-16GB', 'memory_used': '0.0 GiB', 'total_memory': '15.8 GiB'}, {'device_id': 2, 'device_name': 'Tesla V100-SXM2-16GB', 'memory_used': '0.0 GiB', 'total_memory': '15.8 GiB'}, {'device_id': 3, 'device_name': 'Tesla V100-SXM2-16GB', 'memory_used': '0.0 GiB', 'total_memory': '15.8 GiB'}]}

Comment (wanliAlex) [reply src/marqo/tensor_search/tensor_search.py]: leave for future work

**Comment by wanliAlex** [comment]
> Thanks for this PR! Left actionable feedback in the comments.
> 
> Further to dos:
> 
> * Action feedback
> * Create a series of unit tests. One that I propose is to load and unload a model ~10 times, to see if the cache is actually being unloaded (else the computer runs out of memory). Please include testing for unusual and edge cases.
> * Create corresponding Py-Marqo methods and a few unit tests
> * Given this is a  new endpoint, create a few [API tests](https://github.com/marqo-ai/marqo-api-tests/tree/mainline/tests). These don't have to be as exhaustive as the unit tests, but should test every endpoint.
>   
>   * An CUDA API test is proposed: load 2 CUDA models. Unload one. Test to ensure the remaining model is still cached. Test to ensure the unloaded model is indeed unloaded. Create API tests [here](https://github.com/marqo-ai/marqo-api-tests/tree/mainline/tests/api_tests/cuda). CUDA test classes are marked by [this decorator](https://github.com/marqo-ai/marqo-api-tests/blob/b909f35a6de35411340ed427af790d23c5c78643/tests/api_tests/cuda/test_cuda_add_documents.py#L12). This ensures that the API test only runs on the CUDA test environment.

Thanks for the detailed instructions.
Please check the unit test at [repo](https://github.com/marqo-ai/marqo/blob/model-cache-management/tests/tensor_search/test_model_cache_management.py)

Note: for the [edge test](https://github.com/marqo-ai/marqo/blob/2f8f087f3c74dfbc2caaccbe6e4ea5d99122f26e/tests/tensor_search/test_model_cache_management.py#L102), i ran a 50 times iteration and it passed. However, this is time-consuming so I reduce the number of iterations to 10.

**Comment by wanliAlex** [comment]
image


The `model_cache_key` has been updated based on the discussion today.

File: tests/tensor_search/test_model_cache_management.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    def test_edge_case_cuda(self):
+        if self.CUDA_FLAG:
+            test_iterations = 10
+            # Note this is a time consuming test.
+
+            for i in range(test_iterations):
+                eject_model(self.MODEL_1, "cuda")
+                load_model(self.MODEL_1, "cuda")
+
+                for id in range(torch.cuda.device_count()):
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py]: nit: id shadows an inbuilt function in Python. I suggest renaming this to something like `id_`

File: tests/tensor_search/test_model_cache_management.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                    # cuda usage
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py:135]: Can we check that the model shows up as loaded up in `get_loaded_models()` and `get_cuda_info()` when loaded, but not when it is ejected?

File: tests/tensor_search/test_model_cache_management.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+
+def load_model(model_name: str, device: str) -> None:
+    validated_model_properties = _validate_model_properties(model_name, None)
+    model_cache_key = _create_model_cache_key(model_name, device, validated_model_properties)
+    _update_available_models(model_cache_key, model_name, validated_model_properties, device, True)
+
+
+class TestModelCacheManagement(MarqoTestCase):
+
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py:18]: Can we test that a generic model can be loaded, checked (via get...info endpoints) and deloaded, too?

File: tests/tensor_search/test_model_cache_management.py (hunk 7 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py]: done.

File: tests/tensor_search/test_model_cache_management.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+            except ModelNotInCache:
+                pass
+        else:
+            pass
+
+
+    def test_cuda_info(self):
+        try:
+            get_cuda_info()
+        except HardwareCompatabilityError:
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py]: We need to ensure that if the machine has CUDA enabled, that this error is NOT thrown.

File: tests/tensor_search/test_model_cache_management.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+                eject_model(my_test_model_2, "cuda")
+            except ModelNotInCache:
+                pass
+        else:
+            pass
+
+
+    def test_cuda_info(self):
+        try:
+            get_cuda_info()
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py]: Can we do some simple tests about the structure of `get_cuda_info` - for example, making the sure the fields are correct

File: tests/tensor_search/test_model_cache_management.py (hunk 9 above)
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py:18]: Can we test the `get_cpu_info` endpoint too? We need to ensure that the structure is stable for that endpoint

File: tests/tensor_search/test_model_cache_management.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+
+
+    def test_edge_case_cpu(self):
+        test_iterations = 10
+        # Note this is a time consuming test.
+
+        for i in range(test_iterations):
+            eject_model(self.MODEL_1, "cpu")
+            load_model(self.MODEL_1, "cpu")
+            # cpu usage
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py:152]: Can we check that the model is present when loaded, and not present when ejected?

File: tests/tensor_search/test_model_cache_management.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+            eject_model(model_name, "cpu")
+            if (model_name, "cpu") in available_models:
+                raise AssertionError
+
+
+        my_test_model_1 = "test-model-1"
+        my_test_model_2 = "test-model-2"
+
+        try:
+            eject_model(my_test_model_1, "cpu")
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py:57]: Add an assertion error under the eject function, to ensure that the test doesn't pass if `ModelNotInCache` isn't thrown

File: tests/tensor_search/test_model_cache_management.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+        except ModelNotInCache:
+            pass
+
+        try:
+            eject_model(my_test_model_2, "cpu")
```
Comment (pandu-k) [inline tests/tensor_search/test_model_cache_management.py:63]: Add an assertion error under the eject function, to ensure that the test doesn't pass if `ModelNotInCache` isn't thrown

**Review by pandu-k — CHANGES_REQUESTED** [review]

File: tests/tensor_search/test_model_cache_management.py (hunk 10 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py]: done

File: tests/tensor_search/test_model_cache_management.py (hunk 11 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py]: done

File: tests/tensor_search/test_model_cache_management.py (hunk 9 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py:18]: done

File: tests/tensor_search/test_model_cache_management.py (hunk 14 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py:63]: done

File: tests/tensor_search/test_model_cache_management.py (hunk 13 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py:57]: done

File: tests/tensor_search/test_model_cache_management.py (hunk 12 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py:152]: added a new test

File: tests/tensor_search/test_model_cache_management.py (hunk 9 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py:18]: added a new test

File: tests/tensor_search/test_model_cache_management.py (hunk 8 above)
Comment (wanliAlex) [reply tests/tensor_search/test_model_cache_management.py:135]: added a new test

File: src/marqo/tensor_search/tensor_search.py [hunk 15]
```diff
@@ -1330,3 +1331,41 @@ def _get_model_properties(index_info):
...
+def get_cuda_info() -> dict:
+    if torch.cuda.is_available():
+        return {"cuda_devices": [{"device_id" : _device_id, "device_name" : torch.cuda.get_device_name(_device_id),
+                "memory_used":f"{round(torch.cuda.memory_allocated(_device_id) / 1024**3, 1)} GiB",
+                "total_memory": f"{round(torch.cuda.get_device_properties(_device_id).total_memory/ 1024**3, 1)} GiB"}
+                for _device_id in range(torch.cuda.device_count())]}
+
+    else:
+        raise errors.HardwareCompatabilityError(message=str(
+            "ERROR: cuda is not supported in your machine!!"
```
Comment (pandu-k) [inline src/marqo/tensor_search/tensor_search.py:1370]: Can you make the error message stylistically similar to this existing one:

https://github.com/marqo-ai/marqo/blob/590965d0db0805e744529358ae42ab867dcf5215/src/marqo/tensor_search/web/api_validation.py#L67

**Review by pandu-k — APPROVED** [review]
